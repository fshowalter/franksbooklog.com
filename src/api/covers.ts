import { getImage } from "astro:assets";
import fs from "node:fs";
import path from "node:path";
import sharp from "sharp";

import {
  createCacheConfig,
  createCacheKey,
  ensureCacheDir,
  getCachedItem,
  saveCachedItem,
} from "~/utils/cache";

import { normalizeSources } from "./utils/normalizeSources";

export type CoverImageProps = {
  height: number;
  src: string;
  srcSet: string;
  width: number;
};

type Work = {
  slug: string;
};

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/covers/*.png",
);

const cacheConfig = createCacheConfig("cover-base64");

export async function getCoverHeight(coverPath: string, targetWidth: number) {
  try {
    const { height, width } = await sharp(coverPath).metadata();

    return (height / width) * targetWidth;
  } catch (error) {
    console.error("Error:", error);
    return 0;
  }
}

export async function getCoverWidth(work: Work, targetHeight: number) {
  try {
    const coverPath = getWorkCoverPath(work);
    const { height, width } = await sharp(coverPath).metadata();

    return (width / height) * targetHeight;
  } catch (error) {
    console.error("Error:", error);
    return 0;
  }
}

export async function getFeedCoverProps(work: Work): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);

  const optimizedImage = await getImage({
    format: "jpeg",
    height: 750,
    quality: 80,
    src: workCoverFile.default,
    width: 500,
  });

  return {
    height: 750,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: 500,
  };
}

export async function getFixedCoverImageProps(
  work: Work,
  { width }: { width: number },
): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);
  const workCoverPath = getWorkCoverPath(work);

  const height = await getCoverHeight(workCoverPath, width);

  const optimizedImage = await getImage({
    densities: [1, 2],
    format: "avif",
    height,
    quality: 80,
    src: workCoverFile.default,
    width: width,
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: width,
  };
}

export async function getFluidCoverImageProps(
  work: Work,
  { width }: { width: number },
): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);
  const coverFilePath = getWorkCoverPath(work);
  const height = await getCoverHeight(coverFilePath, width);

  const optimizedImage = await getImage({
    format: "avif",
    height,
    quality: 80,
    src: workCoverFile.default,
    width: width,
    widths: [0.25, 0.5, 1, 2].map((w) => w * width),
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width,
  };
}

export async function getOpenGraphCoverAsBase64String(work: Work) {
  const width = 420;
  const format = "png";
  let cacheKey = "";

  if (cacheConfig.enableCache) {
    await ensureCacheDir(cacheConfig.cacheDir);

    const cacheKeyData = `${work.slug}-${width}-${format}`;
    cacheKey = createCacheKey(cacheKeyData);

    const cachedCover = await getCachedItem<string>(
      cacheConfig.cacheDir,
      cacheKey,
      "txt",
      false,
      cacheConfig.debugCache,
      `Cover base64: ${work.slug}`,
    );

    if (cachedCover) {
      return cachedCover;
    }
  }

  const imageBuffer = await sharp(getWorkCoverPath(work))
    .resize(width)
    .toFormat(format)
    .toBuffer();

  const base64String = `data:${"image/png"};base64,${imageBuffer.toString("base64")}`;

  if (cacheConfig.enableCache) {
    await saveCachedItem(cacheConfig.cacheDir, cacheKey, "txt", base64String);
  }

  return base64String;
}

export async function getStructuredDataCoverSrc(work: Work): Promise<string> {
  const workCoverFile = await getWorkCoverFile(work);

  const optimizedImage = await getImage({
    format: "jpeg",
    height: 750,
    quality: 80,
    src: workCoverFile.default,
    width: 500,
  });

  return normalizeSources(optimizedImage.src);
}

export async function getUpdateCoverProps(
  work: Work,
): Promise<CoverImageProps> {
  const coverFile = await getWorkCoverFile(work);

  const coverFilePath = getWorkCoverPath(work);
  const height = await getCoverHeight(coverFilePath, 500);

  const optimizedImage = await getImage({
    format: "png",
    height,
    quality: 100,
    src: coverFile.default,
    width: 500,
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: 500,
  };
}

export function getWorkCoverPath(work: Work) {
  const workCover = coverPath(work.slug);

  if (workCover) {
    return workCover;
  }

  return coverPath("default") || "";
}

function coverPath(slug: string) {
  const coverPath = path.resolve(`./content/assets/covers/${slug}.png`);
  if (fs.existsSync(coverPath)) {
    return coverPath;
  }

  return;
}

async function getWorkCoverFile(work: Work) {
  const coverKey = Object.keys(images).find((image) => {
    return image.endsWith(`${work.slug}.png`);
  });

  if (coverKey) {
    return await images[coverKey]();
  }

  const defaultWorkCoverKey = Object.keys(images).find((image) => {
    return image.endsWith(`default.png`);
  })!;

  return await images[defaultWorkCoverKey]();
}
