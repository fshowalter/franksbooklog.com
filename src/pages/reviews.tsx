import { graphql } from "gatsby";
import { HeadBuilder, Reviews } from "../components";

export function Head(): JSX.Element {
  return (
    <HeadBuilder
      pageTitle="Reviews"
      description="A sortable and filterable list of every book or short story I've read and reviewed since 2022."
      image={null}
      article={false}
    />
  );
}

export default function ReviewsPage({
  data,
}: {
  data: Queries.ReviewsPageQuery;
}): JSX.Element {
  return (
    <Reviews
      items={data.reviewedWork.nodes}
      shortStoryCount={data.shortStory.totalCount}
      bookCount={data.books.totalCount}
      abandonedCount={data.abandoned.totalCount}
      distinctKinds={data.reviewedWork.kinds}
      distinctPublishedYears={data.reviewedWork.publishedYears}
      distinctReviewYears={data.reviewedWork.reviewYears}
      initialSort="author-asc"
    />
  );
}

export const pageQuery = graphql`
  query ReviewsPage {
    books: allReviewsJson(filter: { kind: { ne: "Short Story" } }) {
      totalCount
    }
    shortStory: allReviewsJson(filter: { kind: { eq: "Short Story" } }) {
      totalCount
    }
    abandoned: allReviewsJson(filter: { gradeValue: { eq: 0 } }) {
      totalCount
    }
    reviewedWork: allReviewsJson(sort: { authors: { sortName: ASC } }) {
      nodes {
        ...ReviewsListItem
      }
      publishedYears: distinct(field: { yearPublished: SELECT })
      reviewYears: distinct(field: { yearReviewed: SELECT })
      kinds: distinct(field: { kind: SELECT })
    }
  }
`;
