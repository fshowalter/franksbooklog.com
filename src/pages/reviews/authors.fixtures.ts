export const data = {
  author: {
    nodes: [
      {
        name: "Leif Babin",
        slug: "leif-babin",
        sortName: "Babin, Leif",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "James M. Cain",
        slug: "james-m-cain",
        sortName: "Cain, James M.",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Scott Carney",
        slug: "scott-carney",
        sortName: "Carney, Scott",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Raymond Chandler",
        slug: "raymond-chandler",
        sortName: "Chandler, Raymond",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Dave Draper",
        slug: "dave-draper",
        sortName: "Draper, Dave",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Bret Easton Ellis",
        slug: "bret-easton-ellis",
        sortName: "Ellis, Bret Easton",
        reviewedWorkCount: 2,
        workCount: 2,
        avatar: null,
      },
      {
        name: "Ian Fleming",
        slug: "ian-fleming",
        sortName: "Fleming, Ian",
        reviewedWorkCount: 3,
        workCount: 4,
        avatar: null,
      },
      {
        name: "Philip Fracassi",
        slug: "philip-fracassi",
        sortName: "Fracassi, Philip",
        reviewedWorkCount: 10,
        workCount: 10,
        avatar: null,
      },
      {
        name: "David Goggins",
        slug: "david-goggins",
        sortName: "Goggins, David",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Robert Greene",
        slug: "robert-greene",
        sortName: "Greene, Robert",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Conner Habib",
        slug: "conner-habib",
        sortName: "Habib, Conner",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "David H. Hackworth",
        slug: "david-h-hackworth",
        sortName: "Hackworth, David H.",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Jesse Itzler",
        slug: "jesse-itzler",
        sortName: "Itzler, Jesse",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Joe Kenda",
        slug: "joe-kenda",
        sortName: "Kenda, Joe",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Stephen King",
        slug: "stephen-king",
        sortName: "King, Stephen",
        reviewedWorkCount: 5,
        workCount: 49,
        avatar: {
          childImageSharp: {
            gatsbyImageData: {
              layout: "fixed" as const,
              placeholder: {
                fallback:
                  "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAAsTAAALEwEAmpwYAAAD1klEQVR42k1UySv/URR9QiLCwhxS5igsZFZYKCV2EhmyQKYorMwL87AxZ57neZ5noshQbChTVpI/4Pw6t776LW7v+33v886959xzn+rr68PU1BSGhoYwOTmJ4eFhDAwMYGRkBGtra1hfX5f1+PgYp6en8n95eRkLCwuybm5uYmVlRdbt7W2oubk5LC4uYmJiArOzsxgdHZV1aWkJZWVlSE5ORmZmJtLT09Hc3CzfEnR1dVWCyQh0cHCAvb09KFY1MzMjWfjB4eEh9vf3kZeXB0NDQ1hbW8PR0REWFhZwcHBAWFgYampq5FveIRiD905OTqCYkVkY3CSVqKgo6OnpwdjYGLa2tvD09ISHhwecnZ0F3MnJCbW1tTg6OvqrjmCURBHgf8CkpCS5YGdnB0tLS4SEhKCyslJWc3NzuLu7S/j7+2Nra0sASZVgZ2dnUPPz86IXs7ARfn5+CA0NldXLywsJCQn4+fnB5eUlsrOzZd/X1xempqaiKYEIyGqJodhharizs4Pu7m4YGRkJLVbASlnx19eXxNvbm1B1c3ODvr4+UlNTcXFxIXcJSCcotl4TrJA0dXV1YWJiAjMzM3R2duL391cAv7+/5RvKoaOjg8jISGmgpin8LRWSNpszPT0tXaRGBgYG8PHxEU9yv62tTX739vbCxsYG2traaGhowO7uLjY2NgSMlaqOjg4x8+DgoFBub29HaWkpgoODhQ734uLikJaWhvj4eNl3dXVFQEAAqqqqRH82lsamlRQvcEoI1NTUhKKiItEpPDwcj4+PQpXmJ3XSra+vR0ZGhnxfWFgoU0a5SJvVKlZGwP7+fqFEyjExMQgMDMT5+TleXl6EEs84BDc3N2hsbER1dTV6enr+qiNdOkGxw2NjY5KdWtEqpMgE9/f3AnB3d4fn52fpJK1CYGrKewSjHzmOLEBRA1Ji6cxGm5SXl+P19VUyUkdWyITUqKurSxpDVrzDyug/NkVGj4DsMrPxsKWlRYxbUVEh/xm8SBfU1dWhoKAAJSUlAkpGrIyWYcjojY+Py0tDtz88PEiW6OhoBAUFISUlRZpBBhp7UMvW1lbw2SMjVkYpeC6TQstQH3aTmj09PYl+NG1ERIREYmIicnNzkZ+fL8n4nFF7BsGoHQGly5+fnzJSFJ8HbD91y8rKksuxsbEy2y4uLjI9WlpastIyZMWqNa+OGPv9/V0acHt7K3Svrq7w8fEh3eQzxseBNuJ02Nvby8q3kcFOX19fCyi9yBdLcT5ZJYNGpgV4QIFzcnIE0NvbW2acYHxwWa2VlRWKi4vFUtSRjaXW/wALBD16KdRXqwAAAABJRU5ErkJggg==",
              },
              images: {
                fallback: {
                  src: "/static/b8efcd83f9ce7a0598e4452e55ab31a3/46322/stephen-king.jpg",
                  srcSet:
                    "/static/b8efcd83f9ce7a0598e4452e55ab31a3/46322/stephen-king.jpg 48w,\n/static/b8efcd83f9ce7a0598e4452e55ab31a3/aec58/stephen-king.jpg 96w",
                  sizes: "48px",
                },
                sources: [
                  {
                    srcSet:
                      "/static/b8efcd83f9ce7a0598e4452e55ab31a3/1da62/stephen-king.avif 48w,\n/static/b8efcd83f9ce7a0598e4452e55ab31a3/9b476/stephen-king.avif 96w",
                    type: "image/avif",
                    sizes: "48px",
                  },
                ],
              },
              width: 48,
              height: 48,
            },
          },
        },
      },
      {
        name: "Kathe Koja",
        slug: "kathe-koja",
        sortName: "Koja, Kathe",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Richard Laymon",
        slug: "richard-laymon",
        sortName: "Laymon, Richard",
        reviewedWorkCount: 5,
        workCount: 131,
        avatar: {
          childImageSharp: {
            gatsbyImageData: {
              layout: "fixed" as const,
              placeholder: {
                fallback:
                  "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAAsTAAALEwEAmpwYAAADlklEQVR42n1U2Su3XRR9/gVFZsksY+aEC0qIRDIkFzJcKCUpw43hggtzkl6ZS0IkU0imMmSeMhTJHBfmeVpfa9ejT6/eU6fn1/mdvc7aa6+9lb29PRweHuLi4gJXV1e4v7/H4+Mj3t/fcXNzg7OzMxwdHWFzc1PucD0/P8udp6enH/vh4QHK9fW1gKiHvPz6+oq5uTmUlpaioKAA1dXVyMzMRGpqKhobGyXw5eXlV1CFANzqAcEqKioQExOD+Ph4+cbGxsLb2xuurq7Q0dFBRkaGZPAbU0X9wT+4xsbG4OzsDBsbG/j5+SEyMhJJSUnCkGzz8vIQGBiI4uJiYanG/gDkwdvbG05OTuDv7w9jY2P4+PggOzsbo6OjOD09xd3dnWi8vb0tEgQHB8vjXJTgByA15GptbYWTkxNyc3Oxvr4uD318fIgMvMciEbCrqwvR0dEoLy//BlRl+wFYWVmJkJAQLCwsCCMCqq8zPQYdHx+ju7sboaGhyMrK+jfD+vp6uLi4oL+/X9hQBgLd3t7i8vIS+/v7WF5eRlNTE3x9faVo/wTs6emBra2tAG5sbGBlZQUDAwOS/tDQEAoLC1FTU4OUlBR4eHhI0ehjVlwFVVRDfn5+Yn5+HnFxcairqxO2FJ8WmpycFP81NzdLdcksISEBXl5ewvjr6+snIDdfYUpFRUWora3F8PAw2tvbMTg4iL6+PnR2dqKtrQ2zs7P48+cPkpOTERQU9M1QtY7y/w5h6hMTExgfH8fU1JSkPTIygtXVVQFiVfkAH2RRCHhwcPA7oNolbLnp6Wns7OxgZmZG9GQLRkVFSetRFjJNTExEeHi4eJdxfwGqnltaWpIC8CI3h8Pi4qJIQF0pAzPIz8+XtKm9Woe/APkSJwqD1tbWhCEtQsN3dHRI51RVVUmRCBYQECDZ0F4kQ8kUAqkG5uYiUEtLi2jHzRYjCKvc29uLhoYGaU1tbW0ZFqz61taWxCpqqlx8id6jz2hcpsg5yHl4fn4uFikrK5PhYGRkJGC6urrQ1NSEu7s7SkpKoLBCu7u7khq9ZW9vDy0trW/R09LSpAhMOycnB3Z2dsLMwMDgmyGHibW1NSwsLKCwcjSoubk5TExMYGZmJpdo6PT0dJmBbm5uAqSnpyeMmK4KyDOytbS0lDuKg4MDHB0dhZmVlRVMTU2hoaEhs48e4zciIgKenp4SwGKEhYXJHYLr6+vD0NBQ4hj/H93zu1DnEgclAAAAAElFTkSuQmCC",
              },
              images: {
                fallback: {
                  src: "/static/d1a5b6a02309b16c0c0494baf7674d38/46322/richard-laymon.jpg",
                  srcSet:
                    "/static/d1a5b6a02309b16c0c0494baf7674d38/46322/richard-laymon.jpg 48w,\n/static/d1a5b6a02309b16c0c0494baf7674d38/aec58/richard-laymon.jpg 96w",
                  sizes: "48px",
                },
                sources: [
                  {
                    srcSet:
                      "/static/d1a5b6a02309b16c0c0494baf7674d38/1da62/richard-laymon.avif 48w,\n/static/d1a5b6a02309b16c0c0494baf7674d38/9b476/richard-laymon.avif 96w",
                    type: "image/avif",
                    sizes: "48px",
                  },
                ],
              },
              width: 48,
              height: 48,
            },
          },
        },
      },
      {
        name: "Ross Macdonald",
        slug: "ross-macdonald",
        sortName: "Macdonald, Ross",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Johnny Marr",
        slug: "johnny-marr",
        sortName: "Marr, Johnny",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "John McPhee",
        slug: "john-mcphee",
        sortName: "McPhee, John",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Edmund Morris",
        slug: "edmund-morris",
        sortName: "Morris, Edmund",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Steven Novella",
        slug: "steven-novella",
        sortName: "Novella, Steven",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Chuck Palahniuk",
        slug: "chuck-palahniuk",
        sortName: "Palahniuk, Chuck",
        reviewedWorkCount: 2,
        workCount: 8,
        avatar: null,
      },
      {
        name: "Andrew Roberts",
        slug: "andrew-roberts",
        sortName: "Roberts, Andrew",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Brad Schoenfeld",
        slug: "brad-schoenfeld",
        sortName: "Schoenfeld, Brad",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Arnold Schwarzenegger",
        slug: "arnold-schwarzenegger",
        sortName: "Schwarzenegger, Arnold",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Jerry Seinfeld",
        slug: "jerry-seinfeld",
        sortName: "Seinfeld, Jerry",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Julie Sherman",
        slug: "julie-sherman",
        sortName: "Sherman, Julie",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "John Skipp",
        slug: "john-skipp",
        sortName: "Skipp, John",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Craig Spector",
        slug: "craig-spector",
        sortName: "Spector, Craig",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Bram Stoker",
        slug: "bram-stoker",
        sortName: "Stoker, Bram",
        reviewedWorkCount: 1,
        workCount: 2,
        avatar: null,
      },
      {
        name: "Donna Tartt",
        slug: "donna-tartt",
        sortName: "Tartt, Donna",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Lol Tolhurst",
        slug: "lol-tolhurst",
        sortName: "Tolhurst, Lol",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
      {
        name: "Jocko Willink",
        slug: "jocko-willink",
        sortName: "Willink, Jocko",
        reviewedWorkCount: 1,
        workCount: 1,
        avatar: null,
      },
    ],
  },
};
