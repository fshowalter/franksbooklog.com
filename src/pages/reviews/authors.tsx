import { graphql } from "gatsby";
import { AuthorAvatarListWithFilters } from "../../components/AuthorAvatarListWithFilters";
import { HeadBuilder } from "../../components/HeadBuilder";

export function Head(): JSX.Element {
  return (
    <HeadBuilder
      pageTitle={`Authors`}
      description={`A sortable and filterable list of authors I've read.`}
      image={null}
      article={false}
    />
  );
}

export default function ReviewAuthorsPage({
  data,
}: {
  data: Queries.ReviewsAuthorsPageQuery;
}): JSX.Element {
  return <AuthorAvatarListWithFilters authors={data.author.nodes} />;
}

export const pageQuery = graphql`
  query ReviewsAuthorsPage {
    author: allAuthorsJson(
      filter: { reviewedWorkCount: { gt: 0 } }
      sort: { sortName: ASC }
    ) {
      nodes {
        ...AuthorAvatarListItem
      }
    }
  }
`;
