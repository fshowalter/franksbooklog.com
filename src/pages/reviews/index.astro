---
import React from "react";

import { Backdrop } from "~/components/Backdrop";
import { Layout as ComponentLayout } from "~/components/Layout/Layout";
import { getProps } from "~/components/Reviews/getProps";
import { Reviews } from "~/components/Reviews/Reviews";
import Layout from "~/layouts/Layout.astro";

const props = await getProps();

// Separate layout props from interactive props
const { backdropImageProps, deck, ...interactiveProps } = props;
---

<Layout
  canonical={true}
  meta={{
    description:
      "All my book and story reviews. Filter by title, kind, review date, work date, or grade. Sort by best or worst, oldest or newest, author, or title.",
  }}
  openGraph={{ description: deck, type: "website" }}
  title="Reviews"
>
  <Fragment slot="meta-theme-color">
    <meta content="rgba(0,0,0,.85)" name="theme-color" />
  </Fragment>
  <ComponentLayout className="bg-subtle">
    <Backdrop deck={deck} imageProps={backdropImageProps} title="Reviews" />
    <Reviews client:load {...interactiveProps} />
  </ComponentLayout>
</Layout>
