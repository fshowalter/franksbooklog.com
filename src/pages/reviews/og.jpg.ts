import type { APIRoute } from "astro";

import { getOpenGraphBackdropAsBase64String } from "~/api/backdrops";
import { OpenGraphImage } from "~/components/open-graph-image/OpenGraphImage";
import { componentToImage } from "~/utils/componentToImage";

/**
 * Astro API endpoint that generates the Open Graph image for the reviews index page.
 * Creates a JPEG image with the reviews backdrop and "Reviews" title for social
 * media sharing when the reviews listing page is shared on platforms like Facebook, Twitter, etc.
 * 
 * @returns {Promise<Response>} HTTP response containing the generated JPEG image with appropriate content-type headers
 */
export const GET: APIRoute = async function get() {
  const jpeg = await componentToImage(
    OpenGraphImage({
      backdrop: await getOpenGraphBackdropAsBase64String("reviews"),
      title: "Reviews",
    }),
  );

  return new Response(jpeg, {
    headers: {
      "Content-Type": "image/jpg",
    },
  });
};
