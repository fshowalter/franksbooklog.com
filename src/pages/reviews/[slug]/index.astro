---
import { allReviews, getContentPlainText } from "~/api/reviews";
import { getProps } from "~/components/Review/getProps";
import { Review } from "~/components/Review/Review";
import Layout from "~/layouts/Layout.astro";
import { toSentenceArray } from "~/utils";

interface Props {
  contentPlainText: string;
  slug: string;
}

export async function getStaticPaths() {
  const { reviews } = await allReviews();

  return await Promise.all(
    reviews.map(async (review) => {
      return {
        params: {
          slug: review.slug,
        },
        props: {
          contentPlainText: getContentPlainText(review.rawContent),
          slug: review.slug,
        },
      };
    }),
  );
}

const { contentPlainText, slug } = Astro.props;

const props = await getProps(slug);

//trim the string to the maximum length
var description = contentPlainText
  .replaceAll(/\r?\n|\r/g, " ")
  .slice(0, Math.max(0, 160));

//re-trim if we are in the middle of a word
description = description.slice(
  0,
  Math.max(0, Math.min(description.length, description.lastIndexOf(" "))),
);

let title = `${props.value.title} by ${toSentenceArray(props.value.authors.map((author) => author.name)).join("")}`;

const kind = props.value.kind == "Short Story" ? "story" : "book";

title = `${title} ${kind} review`;
---

<Layout
  canonical={true}
  meta={{ description }}
  openGraph={{ description, type: "article" }}
  title={title}
>
  <Fragment slot="meta-theme-color">
    <meta
      content="#fff"
      media="(prefers-color-scheme: light)"
      name="theme-color"
    />
    <meta
      content="#322f2f"
      media="(prefers-color-scheme: dark)"
      name="theme-color"
    />
  </Fragment>
  <Review {...props} />
</Layout>
