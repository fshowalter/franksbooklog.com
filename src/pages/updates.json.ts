import { getUpdateCoverProps } from "~/api/covers";
import { mostRecentReviews } from "~/api/reviews";

const gradeToStars: Record<string, number> = {
  A: 5,
  "A+": 5,
  "A-": 4.5,
  B: 4,
  "B+": 4,
  "B-": 3.5,
  C: 3,
  "C+": 3,
  "C-": 2.5,
  D: 2,
  "D+": 2,
  "D-": 1.5,
  F: 1,
  "F+": 1,
  "F-": 0.5,
};

export async function GET() {
  const reviews = await mostRecentReviews(6);

  const updateItems = await Promise.all(
    reviews.map(async (review) => {
      const coverProps = await getUpdateCoverProps(review);

      return {
        authors: review.authors.map((author) => author.name),
        date: review.date,
        image: coverProps.src,
        slug: review.slug,
        stars: gradeToStars[review.grade],
        title: review.title,
      };
    }),
  );

  return new Response(JSON.stringify(updateItems));
}
