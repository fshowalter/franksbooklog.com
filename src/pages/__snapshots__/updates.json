[
  {
    "date": "2025-02-15T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fthe-lonely-one-by-richard-laymon.png%3ForigWidth%3D500%26origHeight%3D667%26origFormat%3Dpng&w=500&h=667&q=100&f=png",
    "slug": "the-lonely-one-by-richard-laymon",
    "stars": 3,
    "title": "The Lonely One by Richard Laymon"
  },
  {
    "date": "2024-12-01T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fghost-story-by-peter-straub.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "ghost-story-by-peter-straub",
    "stars": 4,
    "title": "Ghost Story by Peter Straub"
  },
  {
    "date": "2024-11-08T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fgoth-by-lol-tolhurst.png%3ForigWidth%3D500%26origHeight%3D754%26origFormat%3Dpng&w=500&h=754&q=100&f=png",
    "slug": "goth-by-lol-tolhurst",
    "stars": 3,
    "title": "Goth by Lol Tolhurst"
  },
  {
    "date": "2024-08-25T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Frecessional-by-david-mamet.png%3ForigWidth%3D500%26origHeight%3D754%26origFormat%3Dpng&w=500&h=754&q=100&f=png",
    "slug": "recessional-by-david-mamet",
    "stars": 2,
    "title": "Recessional by David Mamet"
  },
  {
    "date": "2024-07-01T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Ffast-forward-by-stephen-morris.png%3ForigWidth%3D500%26origHeight%3D784%26origFormat%3Dpng&w=500&h=784&q=100&f=png",
    "slug": "fast-forward-by-stephen-morris",
    "stars": 4.5,
    "title": "Fast Forward by Stephen Morris"
  },
  {
    "date": "2024-06-30T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fgothic-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "gothic-by-philip-fracassi",
    "stars": 4,
    "title": "Gothic by Philip Fracassi"
  }
]
